File: planning/dialog-service.lua
local ____lualib = require("lualib_bundle")
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__ArrayFilter = ____lualib.__TS__ArrayFilter
local ____exports = {}
local ____content_2Dregistry = require("planning.content-registry")
local getNpcById = ____content_2Dregistry.getNpcById
local getQuestById = ____content_2Dregistry.getQuestById
local listQuests = ____content_2Dregistry.listQuests
local ____quest_2Dlog = require("planning.quest-log")
local isQuestAccepted = ____quest_2Dlog.isQuestAccepted
local listQuestEntries = ____quest_2Dlog.listQuestEntries
local QUEST_OFFERS_BY_NPC_ID = {}
local function buildQuestPromptMap(self)
    local quests = listQuests(nil)
    for ____, quest in ipairs(quests) do
        do
            if quest.offerNpcId == nil then
                goto __continue3
            end
            local bucket = QUEST_OFFERS_BY_NPC_ID[quest.offerNpcId] or ({})
            bucket[#bucket + 1] = {
                kind = "offer",
                questId = quest.id,
                questName = quest.name,
                recommendedLevel = quest.recommendedLevel,
                playerLine = quest.conversationStarter or ("Do you have work related to " .. quest.name) .. "?",
                prompt = (quest.name .. ": ") .. quest.summary
            }
            QUEST_OFFERS_BY_NPC_ID[quest.offerNpcId] = bucket
        end
        ::__continue3::
    end
end
buildQuestPromptMap(nil)
function ____exports.getStandardDialogForNpc(self, npcId)
    local npc = getNpcById(nil, npcId)
    return {unpack(npc.standardDialog)}
end
function ____exports.getQuestDialogPromptsForNpc(self, npcId)
    local prompts = {}
    local offers = QUEST_OFFERS_BY_NPC_ID[npcId] or ({})
    for ____, offer in ipairs(offers) do
        do
            if isQuestAccepted(nil, offer.questId) then
                goto __continue8
            end
            prompts[#prompts + 1] = __TS__ObjectAssign({}, offer)
        end
        ::__continue8::
    end
    local readyToTurnIn = __TS__ArrayFilter(
        listQuestEntries(nil),
        function(____, entry) return entry.turnInNpcId == npcId and entry.status == "ready-to-turn-in" end
    )
    for ____, entry in ipairs(readyToTurnIn) do
        prompts[#prompts + 1] = {
            kind = "turn-in",
            questId = entry.questId,
            questName = entry.questName,
            recommendedLevel = getQuestById(nil, entry.questId).recommendedLevel,
            playerLine = ("I've finished " .. entry.questName) .. ".",
            prompt = "Turn in: " .. entry.questName
        }
    end
    return prompts
end
return ____exports

File: src/planning/dialog-service.ts
import { getNpcById, getQuestById, listQuests } from "./content-registry";
import { isQuestAccepted, listQuestEntries } from "./quest-log";

export type QuestDialogPrompt = {
  kind: "offer" | "turn-in";
  questId: string;
  questName: string;
  recommendedLevel: number;
  playerLine: string;
  prompt: string;
};

export type NpcDialogOption = {
  id: string;
  playerLine: string;
  npcResponse: string;
  questReady: boolean;
};

const QUEST_OFFERS_BY_NPC_ID: Record<string, QuestDialogPrompt[]> = {};

function buildQuestPromptMap(): void {
  const quests = listQuests();

  for (const quest of quests) {
    if (quest.offerNpcId === undefined) {
      continue;
    }

    const bucket = QUEST_OFFERS_BY_NPC_ID[quest.offerNpcId] ?? [];
    bucket.push({
      kind: "offer",
      questId: quest.id,
      questName: quest.name,
      recommendedLevel: quest.recommendedLevel,
      playerLine: quest.conversationStarter ?? `Do you have work related to ${quest.name}?`,
      prompt: `${quest.name}: ${quest.summary}`,
    });
    QUEST_OFFERS_BY_NPC_ID[quest.offerNpcId] = bucket;
  }
}

buildQuestPromptMap();

export function getStandardDialogForNpc(npcId: string): string[] {
  const npc = getNpcById(npcId);
  return [...npc.standardDialog];
}

export function getDialogOptionsForNpc(npcId: string): NpcDialogOption[] {
  const npc = getNpcById(npcId);
  if (npc.dialogOptions === undefined || npc.dialogOptions.length === 0) {
    return [
      {
        id: `${npcId}:default-greeting`,
        playerLine: "How are things around town?",
        npcResponse: npc.standardDialog[0] ?? npc.notes,
        questReady: false,
      },
    ];
  }

  return npc.dialogOptions.map((option) => ({
    id: option.id,
    playerLine: option.playerLine,
    npcResponse: option.npcResponse,
    questReady: option.questReady === true,
  }));
}

export function getQuestDialogPromptsForNpc(npcId: string): QuestDialogPrompt[] {
  const prompts: QuestDialogPrompt[] = [];

  const offers = QUEST_OFFERS_BY_NPC_ID[npcId] ?? [];
  for (const offer of offers) {
    if (isQuestAccepted(offer.questId)) {
      continue;
    }

    prompts.push({ ...offer });
  }

  const readyToTurnIn = listQuestEntries().filter(
    (entry) => entry.turnInNpcId === npcId && entry.status === "ready-to-turn-in",
  );
  for (const entry of readyToTurnIn) {
    prompts.push({
      kind: "turn-in",
      questId: entry.questId,
      questName: entry.questName,
      recommendedLevel: getQuestById(entry.questId).recommendedLevel,
      playerLine: `I've finished ${entry.questName}.`,
      prompt: `Turn in: ${entry.questName}`,
    });
  }

  return prompts;
}
